src/librbd/api/Namespace.cc

```diff
@@ -64,22 +64,36 @@ int Namespace<I>::create(librados::IoCtx& io_ctx, const std::string& name)
     return r;
   }
 
+  int ret_val;
   librados::IoCtx ns_ctx;
   ns_ctx.dup(io_ctx);
   ns_ctx.set_namespace(name);
 
+  r = ns_ctx.create(RBD_TRASH, false);
+  if (r < 0) {
+    lderr(cct) << "failed to create trash: " << cpp_strerror(r)
+               << dendl;
+    goto rollback_ns;
+  }
+
```

Comment:
Another option is to set a bool variable to true if trash was created and have a single cleanup goto which will check the variable to see if the trash needs to be deleted.